### PR TITLE
Dynamodb lookup

### PR DESCRIPTION
**Pull request by syphon7**
I created a way to lookup values from DynamoDb tables. I think most basic lookups will look using this PR, but there are probably a lot of edge cases I haven't thought about.

Hopefully the docs aren't too confusing. I am open to any suggestions to make this easier to use.

### CODE REVIEW

File: docs/lookups.rst [hunk 1]
```diff
@@ -293,6 +293,42 @@ values)
 The region can be omitted (e.g. ``DBUser: ${ssmstore MyDBUser}``), in which
 case ``us-east-1`` will be assumed.
 
+.. _`dynamodb lookup`:
```
Comment (phobologic) [inline docs/lookups.rst]: This should be added to the TOC at the top

File: stacker/lookups/handlers/dynamodb.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    """Get a value from a dynamodb table
+
+    dynamodb field types should be in the following format:
+
+        [<region>:<tablename>@]<primarypartionkey>.<keyvalue>.<keyvalue>...
+
+    Note: The region is optional, and defaults to the environment's
+    `AWS_DEFAULT_REGION` if not specified.
+
+
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: nit: probably worth deleting this white space

File: stacker/lookups/handlers/dynamodb.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+TYPE_NAME = "dynamodb"
+
+
+def handler(value, **kwargs):
+    """Get a value from a dynamodb table
+
+    dynamodb field types should be in the following format:
+
+        [<region>:<tablename>@]<primarypartionkey>.<keyvalue>.<keyvalue>...
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: I think the `[]` shouldn't include the `<tablename>@` right?  That's not optional?

File: stacker/lookups/handlers/dynamodb.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        match = re.search(regex_matcher, key)
+        if match:
+            if match.group(1) in ["M", "S", "N", "L"]:
+                match_val = str(match.group(1))
+                key = key.replace(match.group(0), "")
+                new_keys.append({match_val: key})
+                clean_table_keys.append(key)
+            else:
+                raise ValueError(
+                    'Stacker does not support looking up that datatype')
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: I'd share the `datatype` in question.

File: stacker/lookups/handlers/dynamodb.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+        else:
+            new_keys.append({"S": key})
+            clean_table_keys.append(key)
+
+    projection_expression = _buildProjectionExpression(clean_table_keys)
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: Don't use camelcase here.

File: stacker/lookups/handlers/dynamodb.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            Key={
+                table_lookup: new_keys[0]
+            },
+            ProjectionExpression=projection_expression
+        )
+    except ClientError as e:
+        if e.response['Error']['Code'] == 'ResourceNotFoundException':
+            raise ValueError('Cannot find the dynamodb table: ' + table_name)
+        elif e.response['Error']['Code'] == 'ValidationException':
+            raise ValueError('No dynamo record matched that partition key')
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: You probably want a generic `raise` by itself at the end here.  If there's a ClientError, and it doesn't match the two you have here, we almost surely want to raise it.

File: stacker/lookups/handlers/dynamodb.py [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+    # find and return the key from the dynamo data returned
+    if 'Item' in response:
+        if len(response['Item']) > 0:
+            return (_getValFromDict(response['Item'], new_keys[1:]))
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: camelcase

File: stacker/lookups/handlers/dynamodb.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    for k in keylist:
+        for k1 in k:
+            if nextType is None:
+                data = data[k[k1]]
+            else:
+                temp_dict = data[nextType]
+                data = temp_dict[k[k1]]
+            nextType = k1
+    if nextType == "L":
+        return _convertDDBListToList(data[nextType])
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: camelcase

File: stacker/tests/lookups/handlers/test_dynamodb.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+import boto3
+from stacker.tests.factories import SessionStub
+
+
+class TestDynamoDBHandler(unittest.TestCase):
+    client = boto3.client('dynamodb', region_name='us-east-1')
+
+    def setUp(self):
+        self.stubber = Stubber(self.client)
+        self.get_parameters_response = {'Item':
```
Comment (phobologic) [inline stacker/tests/lookups/handlers/test_dynamodb.py]: Can you clean up the white space here a little:

```
self.get_parameters_response = {
    'Item': {
        'TestMap': {
            'M': {
                'String1': {'S': 'StringVal1'},
                'List1': {
                    'L': [ 
                        {'S': 'ListVal1'},
                        {'S': 'ListVal2'}
                    ]
                    'Number1': {'N': '12345'},
                }
            }
        }
    }
}
```

File: stacker/tests/lookups/handlers/test_dynamodb.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                return_value=SessionStub(client))
+    def test_dynamodb_handler(self, mock_client):
+        self.expected_params = {
+            'TableName': "TestTable",
+            'Key': {
+                "TestKey": {'S': 'TestVal'}
+            },
+            'ProjectionExpression': "TestVal,TestMap,String1"
+        }
+        self.base_lookup_key = "TestTable@TestKey:TestVal.TestMap[M].String1"
```
Comment (phobologic) [inline stacker/tests/lookups/handlers/test_dynamodb.py]: You probably don't need `self.` for any of these variables.

**Comment by syphon7** [comment]
Great feedback -Ill get those changes checked in!

File: stacker/lookups/handlers/dynamodb.py [hunk 11]
```diff
@@ -4,63 +4,45 @@
...
+        table_info, table_keys = value.split('@', 1)
+        if ':' in table_info:
+            region, table_name = table_info.split(':', 1)
         else:
             table_name = table_info
     else:
         raise ValueError('Please make sure to include a tablename')
 
-    if table_name is None:
+    if table_name is '':
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: Maybe it's better to just use:

```
if not table_name:
```

That way it'll fail for any non-existant value, None or ''.

File: stacker/lookups/handlers/dynamodb.py [hunk 12]
```diff
@@ -74,35 +56,100 @@ def handler(value, **kwargs):
...
-            'No dynamo record with those paramters could be found')
+            'The dynamodb record could not be found using the following '
+            'key: {}'.format(new_keys[0]))
+
+
+def _lookup_key_parse(table_keys):
+    """Return the order in which the stacks should be executed.
+
+    Args:
+        - dependencies (dict): a dictionary where each key should be the
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: The `-` at the beginning here isn't part of the standard, can you remove it?  Thanks.

File: stacker/lookups/handlers/dynamodb.py [hunk 13]
```diff
@@ -74,35 +56,100 @@ def handler(value, **kwargs):
...
+
+    return key_dict
+
+
+def _build_projection_expression(clean_table_keys):
+    """Given cleaned up keys, this will return a projection expression for
+    the dynamodb lookup.
+
+    Args:
+        clean_table_keys (dictionary): keys without the data types attached
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: `dictionary` -> `dict`

File: stacker/lookups/handlers/dynamodb.py [hunk 14]
```diff
@@ -74,35 +56,100 @@ def handler(value, **kwargs):
...
+
+    Returns:
+        string: A projection expression for the dynamodb lookup.
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: `string` -> `str`

File: stacker/lookups/handlers/dynamodb.py [hunk 15]
```diff
@@ -74,35 +56,100 @@ def handler(value, **kwargs):
...
 
 
-def _getValFromDict(data, keylist):
-    """This finds the key inside the data returned by dynamodb
+def _get_val_from_ddb_data(data, keylist):
+    """Given a dictionary of dynamodb data (including the datatypes) and a
+    properly structured keylist, it will return the value of the lookup
+
+    Args:
+        - data(dictionary): the raw dynamodb data
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: no `- ` at the beginning, change `dictionary` -> `dict`, also you're missing a space between the name and the type

File: stacker/lookups/handlers/dynamodb.py [hunk 16]
```diff
@@ -111,19 +158,30 @@ def _getValFromDict(data, keylist):
...
 
-def _convertDDBListToList(convlist):
-    """This removes the variable types from the list before passing it to the
-        lookup
+def _convert_ddb_list_to_list(conversion_list):
+    """Given a dynamodb list, it will return a python list without the dynamodb
+        datatypes
+
+    Args:
+        - conversion_list(dictionary): a dynamodb list which includes the
```
Comment (phobologic) [inline stacker/lookups/handlers/dynamodb.py]: See above - I'd go through this and just clean this up everywhere.

**Comment by phobologic** [comment]
Thanks @syphon7 !

**Pull request merged**